Localstack: Force S3 to use `path` style

### The problem
Local stack requires to force S3 path to `path` style, whereas by default Go AWS SDK uses `auto` that actually means `virtual`. 
That behavior causes failures by querying S3-related information over the Localstack. Details can be found:
- [ S3 path styles](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html)
- [related Localstack issue](https://github.com/localstack/localstack/issues/1180)
- failure details are described in the [How to reproduce](#How-to-reproduce)

### Desired solution
It is required to introduce an ability to set the boolean value to [S3ForcePathStyle  in aws.Config](https://docs.aws.amazon.com/sdk-for-go/api/aws/#Config).
That can be done by adding the correspondent connection parameter `s3_force_path_style`.

### The alternative solution
Potentially there is an alternative solution that I tried and it is actually does not work at all. 

AWS provides an ability to specify [`addressing_style` in aws credentials file]( https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html#cli-configure-files-s3):
```
Specifies which addressing style to use. This controls whether the bucket name is in the hostname or is part of the URL. Valid values are: path, virtual, and auto. The default value is auto.

There are two styles of constructing an Amazon S3 endpoint. The first is called virtual and includes the bucket name as part of the hostname. For example: https://bucketname.s3.amazonaws.com. Alternatively, with the path style, you treat the bucket name as if it is a path in the URI; for example, https://s3.amazonaws.com/bucketname. The default value in the CLI is to use auto, which attempts to use the virtual style where it can, but will fall back to path style when required. For example, if your bucket name is not DNS compatible, the bucket name cannot be part of the hostname and must be in the path. With auto, the CLI will detect this condition and automatically switch to path style for you. If you set the addressing style to path, you must then ensure that the AWS Region you configured in the AWS CLI matches the Region of your bucket.
```

I have created an AWS profile:
```
[test]
s3 = 
  addressing_style = path
```

Used this profile in AWS steampipe connection:
```
connection "aws" {
   plugin="aws"
   access_key="default_access_key"
   endpoint_url="http://localhost:4566"
   regions=["sa-east-1"]
   secret_key="default_secret_key"
   profile="test"
```

And got the same error.

### How to reproduce

1. [Run Localstack](https://docs.localstack.cloud/get-started/)
2. Configure AWS connection for steampipe providing `endpoint_url` parameter.
Here is an example:
```
connection "aws" {
   plugin="aws"
   access_key="default_access_key"
   endpoint_url="http://localhost:4566"
   regions=["sa-east-1"]
   secret_key="default_secret_key"
```
3. Execute the following query:
```
select name, arn, region, creation_date as create_time, 
  jsonb_path_query(server_side_encryption_configuration, '$.Rules[0].ApplyServerSideEncryptionByDefault') ->> 'KMSMasterKeyID' as kms_key_id
from aws_s3_bucket
union
select name, arn, region, creation_date as create_time,
  null as kms_key_id
from aws_s3_bucket
where server_side_encryption_configuration is null
```

The expected result: query fails with:
```
Error: RequestError: send request failed
caused by: Get "http://lab-bucket-1.localhost:4566/?encryption=": dial tcp 120.78.181.57:4566: i/o timeout (SQLSTATE HV000)
```

Plugin logs:
```
  2022-06-13 13:10:07.552 UTC [ERROR] steampipe-plugin-aws.plugin: [ERROR] 1655125176492: table 'aws_s3_bucket' failed to get column data, callId 1655125176492: table 'aws_s3_bucket' column 'server_side_encryption_configuration' requires hydrate data from getBucketEncryption, which failed with error RequestError: send request failed
  2022-06-13 13:10:16.072 UTC [ERROR] steampipe-plugin-aws.plugin: [ERROR] 1655125176492: callHydrate getBucketEncryption finished with error: RequestError: send request failed
  2022-06-13 13:10:16.072 UTC [WARN]  steampipe-plugin-aws.plugin: [WARN]  1655125176492: hydrate error chan select error RequestError: send request failed
  2022-06-13 13:10:16.072 UTC [WARN]  steampipe-plugin-aws.plugin: [WARN]  1655125176492: getRow failed with error RequestError: send request failed
  2022-06-13 13:10:16.072 UTC [ERROR] steampipe-plugin-aws.plugin: [ERROR] 1655125176492: table 'aws_s3_bucket' failed to get column data, callId 1655125176492: table 'aws_s3_bucket' column 'server_side_encryption_configuration' requires hydrate data from getBucketEncryption, which failed with error RequestError: send request failed
```  

**The root cause:** 
`getBucketEncryption` uses improper (from the Localstack perspective) url `http://**lab-bucket-2.**localhost:4566/?encryption=' 
instead of `localhost:4566/lab-bucket-2?encryption=`.

@cbruno10 Here is a PR #1082 that  fixes my case, so the following query successfully returns valid results:

```sql
 select name, arn, region, creation_date as create_time, 
  jsonb_path_query(server_side_encryption_configuration, '$.Rules[0].ApplyServerSideEncryptionByDefault') ->> 'KMSMasterKeyID' as kms_key_id
from aws_s3_bucket
union
select name, arn, region, creation_date as create_time,
  null as kms_key_id
from aws_s3_bucket
where server_side_encryption_configuration is null order by region
```

Per  your request I've added more buckets to the different regions:
```
+----------------------------+-----------------------------------------+--------------+---------------------------+------------+
| name                       | arn                                     | region       | create_time               | kms_key_id |
+----------------------------+-----------------------------------------+--------------+---------------------------+------------+
| ca-central-1--lab-bucket-1 | arn:aws:s3:::ca-central-1--lab-bucket-1 | ca-central-1 | 2022-06-14T14:40:41-05:00 | <null>     |
| ca-central-1--lab-bucket-2 | arn:aws:s3:::ca-central-1--lab-bucket-2 | ca-central-1 | 2022-06-14T14:40:42-05:00 | <null>     |
| ca-central-1--lab-bucket-3 | arn:aws:s3:::ca-central-1--lab-bucket-3 | ca-central-1 | 2022-06-14T14:40:42-05:00 | <null>     |
| eu-central-1--lab-bucket-1 | arn:aws:s3:::eu-central-1--lab-bucket-1 | eu-central-1 | 2022-06-14T14:40:43-05:00 | <null>     |
| eu-central-1--lab-bucket-2 | arn:aws:s3:::eu-central-1--lab-bucket-2 | eu-central-1 | 2022-06-14T14:40:43-05:00 | <null>     |
| eu-central-1--lab-bucket-3 | arn:aws:s3:::eu-central-1--lab-bucket-3 | eu-central-1 | 2022-06-14T14:40:43-05:00 | <null>     |
| sa-east-1--lab-bucket-1    | arn:aws:s3:::sa-east-1--lab-bucket-1    | sa-east-1    | 2022-06-14T14:40:32-05:00 | <null>     |
| sa-east-1--lab-bucket-2    | arn:aws:s3:::sa-east-1--lab-bucket-2    | sa-east-1    | 2022-06-14T14:40:33-05:00 | <null>     |
| sa-east-1--lab-bucket-3    | arn:aws:s3:::sa-east-1--lab-bucket-3    | sa-east-1    | 2022-06-14T14:40:34-05:00 | <null>     |
| us-east-1--lab-bucket-1    | arn:aws:s3:::us-east-1--lab-bucket-1    | us-east-1    | 2022-06-14T14:40:38-05:00 | <null>     |
| us-east-1--lab-bucket-2    | arn:aws:s3:::us-east-1--lab-bucket-2    | us-east-1    | 2022-06-14T14:40:39-05:00 | <null>     |
| us-east-1--lab-bucket-3    | arn:aws:s3:::us-east-1--lab-bucket-3    | us-east-1    | 2022-06-14T14:40:39-05:00 | <null>     |
| us-east-2--lab-bucket-1    | arn:aws:s3:::us-east-2--lab-bucket-1    | us-east-2    | 2022-06-14T14:40:39-05:00 | <null>     |
| us-east-2--lab-bucket-2    | arn:aws:s3:::us-east-2--lab-bucket-2    | us-east-2    | 2022-06-14T14:40:40-05:00 | <null>     |
| us-east-2--lab-bucket-3    | arn:aws:s3:::us-east-2--lab-bucket-3    | us-east-2    | 2022-06-14T14:40:40-05:00 | <null>     |
| us-west-1--lab-bucket-1    | arn:aws:s3:::us-west-1--lab-bucket-1    | us-west-1    | 2022-06-14T14:40:35-05:00 | <null>     |
| us-west-1--lab-bucket-2    | arn:aws:s3:::us-west-1--lab-bucket-2    | us-west-1    | 2022-06-14T14:40:36-05:00 | <null>     |
| us-west-1--lab-bucket-3    | arn:aws:s3:::us-west-1--lab-bucket-3    | us-west-1    | 2022-06-14T14:40:37-05:00 | <null>     |
| us-west-2--lab-bucket-1    | arn:aws:s3:::us-west-2--lab-bucket-1    | us-west-2    | 2022-06-14T14:40:37-05:00 | <null>     |
| us-west-2--lab-bucket-2    | arn:aws:s3:::us-west-2--lab-bucket-2    | us-west-2    | 2022-06-14T14:40:37-05:00 | <null>     |
| us-west-2--lab-bucket-3    | arn:aws:s3:::us-west-2--lab-bucket-3    | us-west-2    | 2022-06-14T14:40:38-05:00 | <null>     |
+----------------------------+-----------------------------------------+--------------+---------------------------+------------+

```


**Unfortunately, that  does not fix the last 2 queries you have requested to check:**

### select * from aws_s3_access_point

```sql
Welcome to Steampipe v0.14.6
For more information, type .help
> select * from aws_s3_access_point
Error: RequestError: send request failed
caused by: Get "http://000000000000.localhost:4566/v20180820/accesspoint?maxResults=1000": dial tcp: lookup 000000000000.localhost: no such host (SQLSTATE HV000)
> 
```

Plugin log:
```
2022-06-14 20:37:41.636 UTC [ERROR] steampipe-plugin-aws.plugin: [ERROR] 1655238830825: streamRows error chan select: RequestError: send request failed
2022-06-14 20:37:41.636 UTC [WARN]  steampipe-plugin-aws.plugin: [WARN]  1655238830825: Execute call failed - cancelling pending item in cache
```

### select * from aws_s3_account_settings

```sql
select * from aws_s3_account_settings
Error: RequestError: send request failed
caused by: Get "http://000000000000.localhost:4566/v20180820/configuration/publicAccessBlock": dial tcp: lookup 000000000000.localhost: no such host (SQLSTATE HV000)
```

Plugin log:
```
2022-06-14 20:43:15.883 UTC [ERROR] steampipe-plugin-aws.plugin: [ERROR] 1655239148321: table 'aws_s3_account_settings' failed to get column data, callId 1655239148321: table 'aws_s3_account_settings' column 'block_public_acls' requires hydrate data from getAccountBucketPublicAccessBlock, which failed with error RequestError: send request failed
```

--- a/aws/connection_config.go
+++ b/aws/connection_config.go
@@ -15,6 +15,7 @@ type awsConfig struct {
 	MinErrorRetryDelay    *int     `cty:"min_error_retry_delay"`
 	IgnoreErrorCodes      []string `cty:"ignore_error_codes"`
 	EndpointUrl           *string  `cty:"endpoint_url"`
+	S3ForcePathStyle      *bool    `cty:"s3_force_path_style"`
 }
 
 var ConfigSchema = map[string]*schema.Attribute{
@@ -46,6 +47,9 @@ var ConfigSchema = map[string]*schema.Attribute{
 	},
 	"endpoint_url": {
 		Type: schema.TypeString,
+	},
+	"s3_force_path_style": {
+		Type: schema.TypeBool,
 	},
 }
 

--- a/aws/service.go
+++ b/aws/service.go
@@ -2006,6 +2006,10 @@ func getSessionWithMaxRetries(ctx context.Context, d *plugin.QueryData, region s
 		sessionOptions.Config.Endpoint = aws.String(awsEndpointUrl)
 	}
 
+	if awsConfig.S3ForcePathStyle != nil {
+		sessionOptions.Config.S3ForcePathStyle = awsConfig.S3ForcePathStyle
+	}
+
 	if awsConfig.Profile != nil {
 		sessionOptions.Profile = *awsConfig.Profile
 	}

--- a/config/aws.spc
+++ b/config/aws.spc
@@ -36,7 +36,6 @@ connection "aws" {
 
   # Set to `true` to force S3 requests to use path-style addressing,
   # i.e., `http://s3.amazonaws.com/BUCKET/KEY`. By default, the S3 client
-  # will use virtual hosted bucket addressing when possible
-  # (`http://BUCKET.s3.amazonaws.com/KEY`).
-  #s3_force_path_style = true
+  # will use virtual hosted bucket addressing when possible (`http://BUCKET.s3.amazonaws.com/KEY`).
+  #s3_force_path_style = false
 }

--- a/config/aws.spc
+++ b/config/aws.spc
@@ -33,4 +33,10 @@ connection "aws" {
   # If not set, the default AWS generated endpoint will be used.
   # Can also be set with the AWS_ENDPOINT_URL environment variable.
   #endpoint_url = "http://localhost:4566"
+
+  # Set to `true` to force S3 requests to use path-style addressing,
+  # i.e., `http://s3.amazonaws.com/BUCKET/KEY`. By default, the S3 client
+  # will use virtual hosted bucket addressing when possible
+  # (`http://BUCKET.s3.amazonaws.com/KEY`).
+  #s3_force_path_style = true
 }

--- a/docs/index.md
+++ b/docs/index.md
@@ -105,9 +105,8 @@ connection "aws" {
   
   # Set to `true` to force S3 requests to use path-style addressing,
   # i.e., `http://s3.amazonaws.com/BUCKET/KEY`. By default, the S3 client
-  # will use virtual hosted bucket addressing when possible
-  # (`http://BUCKET.s3.amazonaws.com/KEY`).
-  #s3_force_path_style = true
+  # will use virtual hosted bucket addressing when possible (`http://BUCKET.s3.amazonaws.com/KEY`).
+  #s3_force_path_style = false
 }
 ```
 
@@ -120,7 +119,7 @@ connection "aws" {
 - `regions` - (Optional) List of AWS regions Steampipe will connect to. Can also be set with the `AWS_REGION` or `AWS_DEFAULT_REGION` environment variables, or the region specified in the active profile.
 - `secret_key` - (Optional) AWS secret key. Can also be set with the `AWS_SECRET_ACCESS_KEY` environment variable.
 - `session_token` - (Optional) Session token for validating temporary credentials. Can also be set with the `AWS_SESSION_TOKEN` environment variable.
-- `s3_force_path_style`- (Optional) The only option to force AWS S3 client to use `path-style` addressing. By default, the S3 client will use virtual hosted bucket addressing when possible.
+- `s3_force_path_style`- (Optional) Specifies whether to use path-style addressing, i.e., `https://s3.amazonaws.com/BUCKET/KEY`, or virtual hosted bucket addressing, i.e., `https://BUCKET.s3.amazonaws.com/KEY`. By default, the S3 client will use virtual hosted bucket addressing when possible.
 
 By default, all options are commented out in the default connection, thus Steampipe will resolve your region and credentials using the same mechanism as the AWS CLI (AWS environment variables, default profile, etc).  This provides a quick way to get started with Steampipe, but you will probably want to customize your experience using configuration options for [querying multiple regions](#multi-region-connections), [configuring credentials](#configuring-aws-credentials) from your [AWS Profiles](#aws-profile-credentials), [SSO](#aws-sso-credentials), [aws-vault](#aws-vault-credentials) etc.
 
@@ -371,7 +370,7 @@ connection "aws_account_123456789012" {
 
 ### IAM Access Key Pair Credentials
 
-сThe AWS plugin allows you set static credentials with the `access_key`, `secret_key`, and `session_token` arguments in your connection.
+The AWS plugin allows you set static credentials with the `access_key`, `secret_key`, and `session_token` arguments in your connection.
 
 ```hcl
 connection "aws_account_x" {

--- a/docs/index.md
+++ b/docs/index.md
@@ -102,6 +102,12 @@ connection "aws" {
   # If not set, the default AWS generated endpoint will be used.
   # Can also be set with the AWS_ENDPOINT_URL environment variable.
   #endpoint_url = "http://localhost:4566"
+  
+  # Set to `true` to force S3 requests to use path-style addressing,
+  # i.e., `http://s3.amazonaws.com/BUCKET/KEY`. By default, the S3 client
+  # will use virtual hosted bucket addressing when possible
+  # (`http://BUCKET.s3.amazonaws.com/KEY`).
+  #s3_force_path_style = true
 }
 ```
 
@@ -114,6 +120,7 @@ connection "aws" {
 - `regions` - (Optional) List of AWS regions Steampipe will connect to. Can also be set with the `AWS_REGION` or `AWS_DEFAULT_REGION` environment variables, or the region specified in the active profile.
 - `secret_key` - (Optional) AWS secret key. Can also be set with the `AWS_SECRET_ACCESS_KEY` environment variable.
 - `session_token` - (Optional) Session token for validating temporary credentials. Can also be set with the `AWS_SESSION_TOKEN` environment variable.
+- `s3_force_path_style`- (Optional) The only option to force AWS S3 client to use `path-style` addressing. By default, the S3 client will use virtual hosted bucket addressing when possible.
 
 By default, all options are commented out in the default connection, thus Steampipe will resolve your region and credentials using the same mechanism as the AWS CLI (AWS environment variables, default profile, etc).  This provides a quick way to get started with Steampipe, but you will probably want to customize your experience using configuration options for [querying multiple regions](#multi-region-connections), [configuring credentials](#configuring-aws-credentials) from your [AWS Profiles](#aws-profile-credentials), [SSO](#aws-sso-credentials), [aws-vault](#aws-vault-credentials) etc.
 
@@ -364,7 +371,7 @@ connection "aws_account_123456789012" {
 
 ### IAM Access Key Pair Credentials
 
-The AWS plugin allows you set static credentials with the `access_key`, `secret_key`, and `session_token` arguments in your connection.
+сThe AWS plugin allows you set static credentials with the `access_key`, `secret_key`, and `session_token` arguments in your connection.
 
 ```hcl
 connection "aws_account_x" {

